### PR TITLE
Add support for different record types in ONNX

### PR DESCRIPTION
**Pull request by antimora**
### Checklist

- [x] Confirm that `run-checks` script has been executed.

### Related Issues/PRs

_Provide links to relevant issues and dependent PRs._

### Changes

This adds support for all record types supported by Burn. Also this includes support for embedding and precision.

### Testing

Added new tests and made sure the existing pass using run-checks

### CODE REVIEW

File: burn-import/src/burn/graph.rs [hunk 1]
```diff
@@ -472,17 +583,34 @@ impl<PS: PrecisionSettings> BurnGraph<PS> {
...
+///
+/// PrettyJson, NamedMpk, and NamedMpkGz use the first option and this struct is used for that.
+struct StructMap<'a, PS: PrecisionSettings>(BurnGraphState<'a, PS>);
+
+/// For custom serialization of the graph state in module struct, there are generally
+/// two options:
+/// 1. Serialize the nodes as a map with the node name as the key and the node as the value.
+/// 2. Serialize the nodes as a tuple.
+///
+/// Mpk, and Bincode use the second option and this struct is used for that.
```
Comment (nathanielsimard) [inline burn-import/src/burn/graph.rs:604]: We should explain the reason why those recorders use those options.

**Review by nathanielsimard — CHANGES_REQUESTED** [review]
Some small changes, but overall good job, the extra builder options are really welcomed.

File: burn-import/onnx-tests/tests/record_type_tests.rs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    named_mpk_half,
+    pretty_json,
+    pretty_json_half,
+    named_mpk_gz,
+    named_mpk_gz_half,
+    bincode,
+    bincode_half,
+    bincode_embedded,
+    bincode_embedded_half
+);
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/onnx-tests/tests/record_type_tests.rs:29]: OK. I'll combine with the other macro.

File: burn-import/src/burn/graph.rs [hunk 3]
```diff
@@ -48,37 +69,111 @@ impl<PS: PrecisionSettings> BurnGraph<PS> {
...
+                    !embed_states,
+                    "Embedding states is not supported for PrettyJsonFileRecorder."
+                );
+
+                self.register_record(
+                    out_file,
+                    embed_states,
+                    &format!("burn::record::PrettyJsonFileRecorder::<{precision_ty_str}>"),
+                );
+            }
```
[the remark this answers is not in the document]
Comment (antimora) [reply burn-import/src/burn/graph.rs:115]: I tried but I didn't like the depth and complexity. Lets leave it for now.

File: burn-import/src/burn/graph.rs (hunk 1 above)
Comment (antimora) [reply burn-import/src/burn/graph.rs]: OK. I'll rewrite my doc comments.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**